lib/src/model/analysis/retro_controller.dart

```diff
@@ -0,0 +1,537 @@
+import 'dart:async';
+import 'dart:math';
+
+import 'package:dartchess/dartchess.dart';
+import 'package:fast_immutable_collections/fast_immutable_collections.dart';
+import 'package:flutter/foundation.dart';
+import 'package:freezed_annotation/freezed_annotation.dart';
+import 'package:lichess_mobile/src/model/analysis/analysis_controller.dart';
+import 'package:lichess_mobile/src/model/analysis/common_analysis_state.dart';
+import 'package:lichess_mobile/src/model/common/chess.dart';
+import 'package:lichess_mobile/src/model/common/eval.dart';
+import 'package:lichess_mobile/src/model/common/id.dart';
+import 'package:lichess_mobile/src/model/common/node.dart';
+import 'package:lichess_mobile/src/model/common/service/move_feedback.dart';
+import 'package:lichess_mobile/src/model/common/service/sound_service.dart';
+import 'package:lichess_mobile/src/model/common/uci.dart';
+import 'package:lichess_mobile/src/model/engine/evaluation_mixin.dart';
+import 'package:lichess_mobile/src/model/engine/evaluation_preferences.dart';
+import 'package:lichess_mobile/src/model/engine/evaluation_service.dart';
+import 'package:lichess_mobile/src/model/explorer/opening_explorer_preferences.dart';
+import 'package:lichess_mobile/src/model/explorer/opening_explorer_repository.dart';
+import 'package:lichess_mobile/src/model/game/exported_game.dart';
+import 'package:lichess_mobile/src/model/game/game_repository_providers.dart';
+import 'package:lichess_mobile/src/network/socket.dart';
+import 'package:lichess_mobile/src/view/engine/engine_gauge.dart';
+import 'package:logging/logging.dart';
+import 'package:riverpod_annotation/riverpod_annotation.dart';
+
+part 'retro_controller.freezed.dart';
+part 'retro_controller.g.dart';
+
+typedef RetroOptions = ({GameId id, Side initialSide});
+
+final Logger _logger = Logger('RetroController');
+
+@freezed
+sealed class Mistake with _$Mistake {
+  const Mistake._();
+
+  const factory Mistake({
+    required ViewBranch branch,
+    @Default(IList<UCIMove>.empty()) IList<UCIMove> openingExplorerSolutions,
+  }) = _Mistake;
+
+  ViewBranch get userBranch => branch.children[0];
+  NormalMove get userMove => userBranch.sanMove.move as NormalMove;
+
+  ViewBranch get serverBranch => branch.children[1];
+  NormalMove get serverMove => serverBranch.sanMove.move as NormalMove;
+
+  bool isSolution(RetroCurrentNode node) =>
+      node.position == serverBranch.position ||
+      node.position.isCheckmate ||
+      // Any move that was found in the master's database is also considered a solution.
+      openingExplorerSolutions.contains(node.sanMove!.move.uci);
+}
+
+/// Depth needed to evaluate alternative solution moves.
+///
+/// https://github.com/lichess-org/lila/blob/d29f27d8cbb0e0dac38308c23c63b828028c085f/ui/analyse/src/retrospect/retroCtrl.ts#L151
```

Comment:
Better not put the link to Lila code actually; line number will most probably change; or just put the link to the whole rectroCtrl file.